src/structures/Message.js

```diff
@@ -242,6 +242,9 @@ class Message extends Base {
   patch(data) {
     const clone = this._clone();
     this._edits.unshift(clone);
+    if (this.client.options.messageEditHistoryMaxSize !== -1) {
+      this._edits.splice(this.client.options.messageEditHistoryMaxSize);
+    }
```

Comment:
The length of the array could be checked before calling `Array#unshift()`, as opposed to unshifting the cloned message and then spilcing it after (when the limit is reached). This would also prevent `_clone()` from being called when the limit is reached.
```suggestion
    const { messageEditHistoryMaxSize } = this.client.options;
    if (messageEditHistoryMaxSize === -1 || this._edits.length < messageEditHistoryMaxSize) {
      const clone = this._clone();
      this._edits.unshift(clone);
    }
```

Reply:
clone is unshifted, pushed onto the head of the array. We're slicing from the end of the array. The only time this optimisation would help is when `messageEditHistoryMaxSize` is `0`, which will be pretty rare considering it has some side-effects.

Reply:
It would also help when the `Message#_edits` has reached the length of the set `messageEditHistoryMaxSize`. Currently, when this happens the message will be cloned and unshifted to the start of the array and then the last element would be spliced out (as you mentioned). With the suggested change, `Message#_clone()` => `Message#_edits#unshift()` => `Message#_edits#splice()` wouldn't even be called everytime the message is edited when the limit has been reached. I'm guessing that this would outweigh the cost of an extra comparison, and will most likely for cases where the `messageEditHistoryMaxSize` to a lower number.

Reply:
That would be absolutely breaking behaviour. If that were implemented, you'd then be in a situation where you were holding super old history and ignoring newer edits.

Reply:
Yes... my apologies I somehow overlooked that, I have updated my comment's suggestion so that it actually updates with incoming new edits and pops off the last element after by checking the returned value of [`Array#unshift()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/unshift) (which is the length of the array) exceeds the `messageEditHistoryMaxSize`. Once again, I'm not fully sure whether this would be more efficient for all cases but I do believe that avoiding the `_clone()`, `unshift()` and `splice()` calls when the limit is reached (especially for lower limits) would be more optimal. Feel free to disregard this suggestion if you disagree.

Reply:
I guess there's also a bit of an unintended advantage to using splice vs pop here, in that if a user decided for any reason to update `messageEditHistoryMaxSize` on the fly, the splice implementation would ensure the edit history is capped to the new value, whereas pop may not if the edit history exceeds `messageEditHistoryMaxSize` and the value was decreased.

Imo, this change doesn't provide much benefit and definitely increases the code complexity. However, happy to be correct if folks with more knowledge of JS internals believe this is significantly more optimal instead of calling splice every time.

Reply:
> However, happy to be correct if folks with more knowledge of JS internals believe this is significantly more optimal instead of calling splice every time.

You called, I delivered.

There's a huge overhead on doing `splice` compared to `pop`, the performance differing by a factor of 2.35 on Node.js v14.5.0.

The reasons why are:

- [`pop()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop) has no arguments, it always removes one value from the array and returns it.
- [`splice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) has multiple optional arguments: `array.splice(start[, deleteCount[, item1[, item2[, ...]]]])`.
- [`splice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) can remove *and* insert, given an amount of elements to remove and the index, which can introduce extra cost.
- [`splice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) creates an array with all the removed elements, which needs it to iterate for each value in it.

The specification can be used to see how much convoluted splice() is compared to pop():

- [`Array.prototype.pop()`](https://tc39.es/ecma262/#sec-array.prototype.pop)
- [`Array.prototype.splice(start, deleteCount, ...items)`](https://tc39.es/ecma262/#sec-array.prototype.splice)

tl;dr: the former is 13 lines in the spec, the latter is 62 lines, with multiple loops and conditionals.

> **Note**: Loading the page might take a while, I suggest opening pop and then scrolling down.

Reply:
Honestly, I'd go for `pop()`, it reads better as in this context we give it the meaning that "we're pushing a new value at the start and removing the last", similar to a FILO queue. I didn't even know `splice(maxNumberOfElements)` was a valid overload, and its behaviour is not obvious for everyone.

> I guess there's also a bit of an unintended advantage to using splice vs pop here, in that if a user decided for any reason to update `messageEditHistoryMaxSize` on the fly, the splice implementation would ensure the edit history is capped to the new value, whereas pop may not if the edit history exceeds `messageEditHistoryMaxSize` and the value was decreased.

Changing the option doesn't immediately trim all the message caches either, but it will for new messages.

Reply:
👍 TIL there was such a performance difference. pop does indeed make sense then. Will change :)